core/shared/src/main/scala/io/circe/Json.scala

```diff
@@ -124,6 +125,27 @@ sealed abstract class Json extends Product with Serializable {
   final def spaces4: String = Printer.spaces4.pretty(this)
 
   /**
+    * Perform a deep merge of this JSON value with another JSON value.
+    *
+    * Objects are merged by key, values from the argument JSON take
+    * precedence over values from this JSON. Nested objects are
+    * recursed.
+    *
+    * Null, Array, Boolean, String and Number are treated as values,
+    * and values from the argument JSON completely replace values
+    * from this JSON.
+    */
+  def deepmerge(that: Json): Json =
+    Apply[Option].map2(asObject, that.asObject) { (lhs, rhs) =>
```

Comment:
I'd personally prefer using either pattern matching or a `for`-comprehension (or just `flatMap`, etc.) here, for a couple of reasons.

The first reason is that this version is probably less clear for most Scala developers. The fact that Cats provides an `Apply` instance for `Option` is great when you need to abstract over type constructors, but here we're just working with plain old `Option`, so I'm not sure `map2` is worth the readability cost.

The second reason is that using the `Apply` instance has additional costs at both compile time and runtime. This is unlikely ever to be relevant in this case specifically, but if we did this everywhere in the codebase it might start to be a problem, and I'd like to aim for consistency. Again, if we needed the abstraction, it'd totally be worth the performance cost in this context, but it doesn't really buy us anything here.